### PR TITLE
fix(a2a): stop intro storms, add intro short-circuit and RPC bearer auth

### PR DESCRIPTION
**Pull request by rileyseaburg**
## Problem

The A2A peer mesh had three issues surfaced by runtime evidence (2,459
of 3,085 sessions on this host were auto-intro junk, 106MB on disk).

1. **LLM-burning intros.** Inbound `message/send` containing the
   auto-intro template was indistinguishable from a real task and was
   executed by `prompt_runtime::run`, costing a completion and a
   persisted session.
2. **Re-intro on every restart.** Discovery dedup was per-process
   in-memory, and the dedup key included the peer name which embeds
   the PID. Every peer restart produced a new identity → re-introduction.
3. **Unauthenticated RPC.** `A2AServer::router()` mounted with no
   `AuthLayer`. Anyone on the LAN could inject prompts into a
   full-tool-access LLM session over `POST /` (the main HTTP server has
   mandatory auth; the A2A peer router did not).

## Fix

- **`src/a2a/intro/`** — new module split into `detect`, `ledger`,
  `send`, `reply`. Intros now carry `metadata[codetether.intro] = true`
  and a legacy-text fallback recognizes the old template.
- **`server.rs`** short-circuits intros in both `message/send` and
  `message/stream` via `server_intro::handle_intro`: canned ack, no
  session, no LLM call, no `session.save()`.
- **`src/a2a/intro/ledger.rs`** persists introduced endpoints to
  `<data_dir>/a2a/introduced.json`. Cross-restart dedup.
- **`spawn.rs`** discovery dedup key changed to `endpoint` only (no
  PID-churned name component).
- **`src/a2a/server_auth.rs`** Bearer-token middleware; opt-in via
  `CODETETHER_AUTH_TOKEN`. Agent-card discovery paths stay public.
- **`server_telemetry.rs`** split out so `server.rs` net-shrinks under
  the 50-line ratchet.

## Evidence

| Claim | Level | Where |
|---|---|---|
| Short-circuit skips LLM | focused local e2e | `tests/a2a_intro_e2e.rs` 2/2 pass with no provider configured |
| Legacy peer text still recognized | focused local e2e | same test |
| Bearer auth works (401/200/public) | focused local e2e | `tests/a2a_auth_e2e.rs` 1/1 pass |
| 4 intro unit tests | focused local | `cargo test --lib a2a::intro` |
| 57 a2a unit tests | focused local | `cargo test --lib a2a` |
| 2,399 junk sessions GC'd | local | moved to `.codetether-agent/backups/intro-gc/` (9.7MB) |
| Full build/type coverage | pending CI | pushed on push to this branch |
| 50-line ratchet | static/local | `./check_file_limits.sh` OK |
| Clippy on changed modules | static/local | clean |

## Compatibility

- Old peers keep working: legacy text match is a second path.
  Receivers short-circuit; legacy senders don't get a normal reply
  flow but also no longer cost LLM time here.
- Mesh stays open by default (no token = no auth) to preserve the
  zero-config experience. Set `CODETETHER_AUTH_TOKEN` to enable.

## Follow-ups (out of scope)

- Redeploy peer at 192.168.50.101 to gain the ledger and stop
  re-introducing per restart.
- Default `SpawnOptions::hostname` could be tightened from `0.0.0.0`
  to loopback when no `--peer` is given — security hardening proposal.

## Note on landing

The two commits were originally pushed directly to `main` from a
local sandbox; this PR re-lands them as `dd86e87` + `6ccabbd` on
`fix/a2a-intro-storms` for review. Once merged, `main` will fast-forward.

### CODE REVIEW

File: src/a2a/intro/ledger.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            }
+        }
+        Err(e) => tracing::debug!(error = %e, "Failed to serialize intro ledger"),
+    }
+}
+
+/// Whether `endpoint` was already introduced to (normalized, no trailing `/`).
+pub fn contains(endpoint: &str) -> bool {
+    load().contains(endpoint.trim_end_matches('/'))
+}
```
[the remark this answers is not in the document]
Comment (rileyseaburg) [reply src/a2a/intro/ledger.rs]: Addressed in dc92be6. `load()` now returns `Result<HashSet<String>, io::Error>` and distinguishes `NotFound` from real I/O errors. `record()` aborts if `load()` errors so a transient read failure cannot clobber a valid ledger. Writes go through a temp-file `rename` for crash safety. The ledger orchestrator was split into 4 child modules (path/load/atomic/record) so each stays under the 50-line ratchet. Added 4 unit tests covering load-missing, contains, record-then-contains round-trip; they use a tempdir for `CODETETHER_DATA_DIR` and a static mutex to serialize parallel runs (env var is process-global).

File: src/a2a/server_auth.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    response::Response,
+};
+const PUBLIC_PATHS: [&str; 2] = ["/.well-known/agent.json", "/.well-known/agent-card.json"];
+pub async fn require_a2a_auth(request: Request<Body>, next: Next) -> Result<Response, StatusCode> {
+    let Ok(expected) = std::env::var("CODETETHER_AUTH_TOKEN") else {
+        return Ok(next.run(request).await);
+    };
+    if expected.is_empty() || PUBLIC_PATHS.contains(&request.uri().path()) {
+        return Ok(next.run(request).await);
+    }
```
[the remark this answers is not in the document]
Comment (rileyseaburg) [reply src/a2a/server_auth.rs:41]: Addressed in dc92be6. The token is now cached in a `OnceLock<Option<String>>` initialized once on first request via `get_or_init`. Subsequent requests hit the process-local cache, so the per-request `env::var` global lock is taken at most once for the lifetime of the process.

File: src/a2a/server_auth.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        .and_then(|v| v.strip_prefix("Bearer "))
+        .ok_or(StatusCode::UNAUTHORIZED)?;
+    if !constant_time_eq(provided.as_bytes(), expected.as_bytes()) {
+        return Err(StatusCode::UNAUTHORIZED);
+    }
+    Ok(next.run(request).await)
+}
+fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
+    a.len() == b.len() && a.iter().zip(b).fold(0u8, |acc, (x, y)| acc | (x ^ y)) == 0
+}
```
[the remark this answers is not in the document]
Comment (rileyseaburg) [reply src/a2a/server_auth.rs]: Good catch. Addressed in dc92be6. `constant_time_eq` is now length-padded: it always walks `a.len().max(b.len()`\) with zero-fill for the shorter side, then ORs the length difference into the result. A length mismatch no longer shortcuts the comparison so token length is not leaked via timing. Function is `pub` and a test exercises all four cases (equal, differ-same-len, prefix, empty).

File: src/a2a/intro/ledger.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+//! Persistent ledger of peers already introduced to.
+//!
+//! The in-memory `discovered` set in the discovery loops forgets on
+//! restart, which caused thousands of duplicate intros. This ledger
+//! persists introduced endpoints to `<data_dir>/a2a/introduced.json`
+//! keyed by endpooint only (peer names embed PIDs and churn per restart).
```
[the remark this answers is not in the document]
Comment (rileyseaburg) [reply src/a2a/intro/ledger.rs]: Fixed in dc92be6. The `endpooint` typo is gone (the entire docs section was rewritten as part of the error-propagating `load()` refactor).

File: tests/a2a_auth_e2e.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    let r = client
+        .post(&url)
+        .bearer_auth("test-secret")
+        .json(&rpc_body)
+        .send()
+        .await
+        .unwrap();
+    assert_eq!(r.status(), 200, "valid bearer must pass auth");
+
+    // 4. Agent card stays public without a token.
```
[the remark this answers is not in the document]
Comment (rileyseaburg) [reply tests/a2a_auth_e2e.rs]: Disagree — the comment as written is accurate. The test does `unsafe set_var(CO...AUTH_TOKEN, "test-secret")` at the top, so `cached_token()` returns `Some("test-secret")` and the middleware enforces auth. Step 4 then hits the `/.well-known/agent.json` path *with* auth enabled and asserts 200. The comment 'stays public even with auth enabled' is exactly the contract being verified. Happy to add a sister test that boots a second server with no token set if you want a positive control alongside the negative-control-with-exception pattern.

File: tests/a2a_intro_e2e.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
+//! End-to-end tests for A2A intro short-circuit and RPC bearer auth.
```
[the remark this answers is not in the document]
Comment (rileyseaburg) [reply tests/a2a_intro_e2e.rs]: Narrowed in 9667713. The module doc now reads 'End-to-end tests for the A2A intro short-circuit path.' Bearer auth is named in the file header only of `a2a_auth_e2e.rs`.
